### PR TITLE
Merge CSI plugin code.

### LINKED ISSUE #7408: Merge csi plugin into velero repo
With CSI-snapshot being more and more commonly used, we should consider merging the repo https://github.com/vmware-tanzu/velero-plugin-for-csi into https://github.com/vmware-tanzu/velero such that users will not need to install the csi plugin when installing velero.

This may save some release overhead and we may still turn on/off the feature by using the feature flag `EnableCSI`

### PR DESCRIPTION
**Pull request by blackpiglet**
Thank you for contributing to Velero!

# Please add a summary of your change
This PR contains many file changes.
To reduce the reviewers' burden, highlight the main change here:
* Replace the external-snapshotter client and the k8s native client with the controller-runtime client.
* Modify how the plugins are registered.
* Refactor some functions.

P.S. Reviewers can ignore the change under the directory `github.com/vmware-tanzu/velero/pkg/restore/actions` because correcting the files' package name is the only change.

# Does your change fix a particular issue?

Fixes #7408 

# Please indicate you've done the following:

- [x] [Accepted the DCO](https://velero.io/docs/v1.5/code-standards/#dco-sign-off). Commits without the DCO will delay acceptance.
- [ ] [Created a changelog file](https://velero.io/docs/v1.5/code-standards/#adding-a-changelog) or added `/kind changelog-not-required` as a comment on this pull request.
- [ ] Updated the corresponding documentation in `site/content/docs/main`.

### CODE REVIEW

**Comment by reasonerjt** [comment]
There are many RIA moved into `restore/actions` but only CSI related BIA into `bakcup/actions`.

For the first step, I wish to suggest put all CSI plugins into `backup/csi` or just into `backup` package and name the file explicitly like `csi_xxxx.go`

**Comment by blackpiglet** [comment]
> There are many RIA moved into `restore/actions` but only CSI related BIA into `bakcup/actions`.
> 
> For the first step, I wish to suggest put all CSI plugins into `backup/csi` or just into `backup` package and name the file explicitly like `csi_xxxx.go`

Backup-related actions are already put into the `github.com/vmware-tanzu/velero/pkg/backup/actions/` in the previous PR.
This PR modifies the restore-related actions because the package names were not set correctly.

File: internal/delete/actions/csi_volumesnapshot_action.go [hunk 1]
```diff
@@ -0,0 +1,119 @@
+/*
+Copyright 2020 the Velero contributors.
```
Comment (Lyndon-Li) [inline internal/delete/actions/csi_volumesnapshot_action.go:2]: Should this way better --- we add a `csi` subfolder, and put all the CSI plugin files there and remove the `csi_` prefix from the files?

File: pkg/util/csi/volume_snapshot.go [hunk 2]
```diff
@@ -242,3 +296,597 @@ func patchVSC(ctx context.Context, snapshotClient snapshotter.SnapshotV1Interfac
...
+func Contains(slice []string, key string) bool {
+	for _, i := range slice {
+		if i == key {
+			return true
+		}
+	}
+	return false
+}
+
+func IsPVCDefaultToFSBackup(pvcNamespace, pvcName string, crClient crclient.Client, defaultVolumesToFsBackup bool) (bool, error) {
```
Comment (Lyndon-Li) [inline pkg/util/csi/volume_snapshot.go:362]: This one looks like a common function, not only for CSI, can we move it to `podvolume` package?

Comment (blackpiglet) [reply pkg/util/csi/volume_snapshot.go:362]: Sure.

File: internal/delete/actions/csi_volumesnapshot_action.go (hunk 1 above)
Comment (blackpiglet) [reply internal/delete/actions/csi_volumesnapshot_action.go:2]: How about deleting the csi prefix?

Comment (blackpiglet) [reply internal/delete/actions/csi_volumesnapshot_action.go:2]: Modified.

File: pkg/util/csi/volume_snapshot.go [hunk 3]
```diff
@@ -242,3 +296,597 @@ func patchVSC(ctx context.Context, snapshotClient snapshotter.SnapshotV1Interfac
...
+// IsVolumeSnapshotHasVSCDeleteSecret returns whether a volumesnapshot should set the deletesnapshot secret
+// for the static volumesnapshotcontent that is created on restore
+func IsVolumeSnapshotHasVSCDeleteSecret(vs *snapshotv1api.VolumeSnapshot) bool {
+	_, nameExists := vs.Annotations[velerov1api.DeleteSecretNameAnnotation]
+	_, nsExists := vs.Annotations[velerov1api.DeleteSecretNamespaceAnnotation]
+	return nameExists && nsExists
+}
+
+// AddAnnotations adds the supplied key-values to the annotations on the object
+func AddAnnotations(o *metav1.ObjectMeta, vals map[string]string) {
```
[the remark this answers is not in the document]
Comment (blackpiglet) [reply pkg/util/csi/volume_snapshot.go:540]: Moved AddAnnotations and AddLabels into the `github.com/vmware-tanzu/velero/pkg/util/util.go`

File: pkg/util/csi/volume_snapshot.go [hunk 4]
```diff
@@ -242,3 +296,597 @@ func patchVSC(ctx context.Context, snapshotClient snapshotter.SnapshotV1Interfac
...
+	if o.Annotations == nil {
+		o.Annotations = make(map[string]string)
+	}
+	for k, v := range vals {
+		o.Annotations[k] = v
+	}
+}
+
+// AddLabels adds the supplied key-values to the labels on the object
+func AddLabels(o *metav1.ObjectMeta, vals map[string]string) {
```
[the remark this answers is not in the document]
Comment (blackpiglet) [reply pkg/util/csi/volume_snapshot.go:550]: Moved AddAnnotations and AddLabels into the github.com/vmware-tanzu/velero/pkg/util/util.go

File: pkg/util/csi/volume_snapshot.go [hunk 5]
```diff
@@ -242,3 +296,597 @@ func patchVSC(ctx context.Context, snapshotClient snapshotter.SnapshotV1Interfac
...
+		return err
+	}
+
+	originVSC := vsc.DeepCopy()
+	vsc.Spec.DeletionPolicy = snapshotv1api.VolumeSnapshotContentDelete
+
+	return crClient.Patch(context.TODO(), vsc, crclient.MergeFrom(originVSC))
+}
+
+func HasBackupLabel(o *metav1.ObjectMeta, backupName string) bool {
```
[the remark this answers is not in the document]
Comment (blackpiglet) [reply pkg/util/csi/volume_snapshot.go:590]: Moved into the `github.com/vmware-tanzu/velero/pkg/util/util.go`.

File: pkg/util/csi/volume_snapshot.go [hunk 6]
```diff
@@ -242,3 +296,597 @@ func patchVSC(ctx context.Context, snapshotClient snapshotter.SnapshotV1Interfac
...
+			if v.PersistentVolumeClaim != nil && v.PersistentVolumeClaim.ClaimName == pvcName {
+				podsUsingPVC = append(podsUsingPVC, p)
+			}
+		}
+	}
+
+	return podsUsingPVC, nil
+}
+
+func GetPodVolumeNameForPVC(pod corev1api.Pod, pvcName string) (string, error) {
```
[the remark this answers is not in the document]
Comment (blackpiglet) [reply pkg/util/csi/volume_snapshot.go:344]: Modified.

File: pkg/util/csi/volume_snapshot.go [hunk 7]
```diff
@@ -242,3 +296,597 @@ func patchVSC(ctx context.Context, snapshotClient snapshotter.SnapshotV1Interfac
 
 	return patched, nil
 }
+
+func GetPVForPVC(
```
[the remark this answers is not in the document]
Comment (blackpiglet) [reply pkg/util/csi/volume_snapshot.go:299]: Modified.

**Review by Lyndon-Li — DISMISSED** [review]

**Review by reasonerjt — DISMISSED** [review]

**Review by Lyndon-Li — APPROVED** [review]

File: pkg/util/csi/volume_snapshot.go [hunk 8]
```diff
@@ -47,59 +48,89 @@ const (
...
+						volumeSnapshotNS, volumeSnapshot),
+				)
+			}
 
-		if tmpVS.Status == nil {
-			return false, nil
-		}
+			if tmpVS.Status == nil {
+				return false, nil
+			}
```
Comment (qiuming-best) [inline pkg/util/csi/volume_snapshot.go:81]: Not sure whether it has an indent problem?

**Review by qiuming-best — APPROVED** [review]

File: pkg/util/csi/volume_snapshot.go (hunk 8 above)
Comment (blackpiglet) [reply pkg/util/csi/volume_snapshot.go:81]: No. This is caused by how the function `wait.PollUntilContextTimeout` is called.

Before modification, the function's parameters are placed into one line.
`err := wait.PollUntilContextTimeout(ctx, waitInternal, timeout, true, func(ctx context.Context) (bool, error) {`

After modification, the functions' parameters are split into different lines.
``` go
	err := wait.PollUntilContextTimeout(
		ctx,
		waitInternal,
		timeout,
		true,
		func(ctx context.Context) (bool, error) {
			tmpVS, err := snapshotClient.VolumeSnapshots(volumeSnapshotNS).Get(
				ctx, volumeSnapshot, metav1.GetOptions{})
```

That caused the intent change.

**Review by reasonerjt — APPROVED** [review]

**Pull request merged**